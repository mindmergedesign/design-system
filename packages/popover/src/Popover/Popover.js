import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import { Manager, Reference, Popper } from 'react-popper'
import { Box, theme } from 'pcln-design-system'
import styled from 'styled-components'
import DEFAULTS_MODIFIERS from '../helpers/defaultModifiers'
import Overlay from '../Overlay'

const defaultProps = {
  theme: theme,
  p: 2,
  bg: 'white',
  borderColor: 'borderGray',
  placement: 'top',
  zIndex: 102,
  width: 400
}

class Popover extends Component {
  constructor(props) {
    super(props)
    this.state = {
      isPopoverOpen: false
    }

    this.contentRef = React.createRef()
    this.triggerRef = React.createRef()
    this.handleToggle = this.handleToggle.bind(this)
    this.handleOpen = this.handleOpen.bind(this)
    this.handleClose = this.handleClose.bind(this)
    this.setFocusToRef = this.setFocusToRef.bind(this)
    this.calcOffset - this.calcOffset.bind(this)
  }

  handleToggle(isOpen) {
    if (isOpen) {
      this.handleClose()
    } else {
      this.handleOpen()
    }
  }

  handleClose() {
    this.setState(
      prevState => {
        if (prevState.isPopoverOpen) {
          return {
            isPopoverOpen: false
          }
        }
      },
      () => {
        this.setFocusToRef(this.triggerRef)
      }
    )
  }

  handleOpen() {
    this.setState(
      prevState => {
        if (!prevState.isPopoverOpen) {
          return {
            isPopoverOpen: true
          }
        }
      },
      () => {
        this.setFocusToRef(this.contentRef)
      }
    )
  }

  setFocusToRef(ref) {
    if (ref) {
      try {
        ref.current.focus()
      } catch {
        return false
      }
    }
  }

  calcOffset(placement) {
    // Need to account for the padding added around the popover, this fixes the offset at the start and end position
    if (RegExp('start*').test(placement)) {
      return '-16px'
    } else if (RegExp('end*').test(placement)) {
      return '16px'
    } else {
      return 0
    }
  }

  render() {
    const styleProps = {
      theme: this.props.theme,
      background: this.props.bg,
      borderColor: this.props.borderColor,
      zIndex: this.props.zIndex,
      width: this.props.width
    }

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            // Need to be a native element, because of ref forwarding limitations with DS functional components
            <div ref={ref}>
              {// Clone element to pass down toggle event so it can be used directly from children as needed
              React.cloneElement(this.props.children, {
                'aria-label': 'Click to open popover with more information',
                onClick: () => this.handleToggle(this.state.isPopoverOpen),
                ref: this.triggerRef //Currently ref only works with native element, if we use a DS core component it does not work.
              })}
            </div>
          )}
        </Reference>
        {this.state.isPopoverOpen &&
          ReactDOM.createPortal(
            <React.Fragment>
              <Popper
                positionFixed={true}
                modifiers={{
                  ...DEFAULTS_MODIFIERS,
                  offset: {
                    offset: this.calcOffset(this.props.placement)
                  }
                }}
                {...this.props}
              >
                {({ placement, ref, style, arrowProps }) => (
                  // Need to be a native element, because of ref forwarding limitations with DS functional components
                  <PopperGuide
                    /*
                     * NOTE: InnerRef has been depracted in V4 of styled components. Make sure to change this prop once we upgrade to styled components v4
                     * https://www.styled-components.com/docs/api#deprecated-innerref-prop
                     */
                    innerRef={ref}
                    style={style}
                    data-placement={placement}
                    aria-label={this.props.ariaLabel || 'Dialog Title'}
                    {...styleProps}
                    role="dialog"
                    describedby={`dialog-description-${this.props.idx}`}
                  >
                    <ContentContainer
                      innerRef={this.contentRef}
                      {...styleProps}
                      tabIndex="-1"
                    >
                      <Box id={`popover-description-${this.props.idx}`}>
                        {this.props.renderContent({
                          handleClose: this.handleClose
                        })}
                      </Box>
                    </ContentContainer>
                    <Arrow
                      innerRef={arrowProps.ref}
                      style={arrowProps.style}
                      data-placement={placement}
                      {...styleProps}
                      aria-hidden="true"
                    />
                  </PopperGuide>
                )}
              </Popper>
              <Overlay
                handleClick={this.handleClose}
                zIndex={this.props.zIndex - 1}
              />
            </React.Fragment>,
            // Append each instance of the Popover as portal directly to the body
            document.querySelector('body')
          )}
      </Manager>
    )
  }
}

const PopperGuide = styled.div`
  padding: 16px;
  z-index: ${({ zIndex }) => zIndex}
  max-width: ${({ width }) => width}px;
  width: 100%;
  box-sizing: border-box;
`
const ContentContainer = styled.section`
  box-shadow: 0 0 0 1px ${({ theme, borderColor }) => theme.colors[borderColor]},
    0 0 4px 0 rgba(0, 0, 0, 0.08), 0 8px 8px 0 rgba(0, 0, 0, 0.08),
    0 16px 16px 0 rgba(0, 0, 0, 0.08);
  font-size: ${({ theme }) => theme.fontSizes[0]}px;
  border-radius: ${({ theme }) => theme.radii[1]}px;
  background: ${({ theme, background }) => theme.colors[background]};
  box-sizing: border-box;
  outline: 0;
  max-width: 100%;
`

// TODO: Create a separate component for Arrow
const ArrowAlignment = () =>
  `
    &[data-placement*="right"] {
      left: 0;
    }
    &[data-placement*="left"] {
      right: 0;
    }
    &[data-placement*="top"] {
      bottom: 0;
    }
    &[data-placement*="bottom"] {
      top: 0;
    }
    &[data-placement*="bottom"]::before,
    &[data-placement*="right"]::before,
    &[data-placement*="top"]::after,
    &[data-placement*="left"]::after {
      content: '';
      position: absolute;
      width: 0;
      height: 0;
      margin: auto;
      border-style: solid;
      transform-origin: 0 0;
      transform: rotate(-45deg);
      border-width: 6px;
    }
  `

const ArrowPlacement = ({ theme, background, borderColor }) => {
  const bgColor = theme.colors[background]
  const brColor = theme.colors[borderColor]

  return `
    &[data-placement*="right"]::before {
      left: 8px;
      border-color: ${bgColor} transparent transparent ${bgColor};
      box-shadow: -0.75px -0.75px 0px 0.25px ${brColor};
    }
    &[data-placement*="left"]::after {
      right: 13px;
      border-color: transparent ${bgColor} ${bgColor} transparent;
      box-shadow: 0.75px 0.75px 0px 0.25px ${brColor};
    }
    &[data-placement*="top"]::after {
      top: -16px;
      margin-left: -5px;
      border-color: transparent transparent ${bgColor} ${bgColor};
      box-shadow: -0.75px 0.75px 0px 0.25px ${brColor};
    }
    &[data-placement*="bottom"]::before {
      top: 16px;
      margin-left: -5px;
      border-color: ${bgColor} ${bgColor} transparent transparent;
      box-shadow: 0.75px -0.75px 0px 0.25px ${brColor};
    }
  `
}
const Arrow = styled.span`
  position: absolute;
  ${ArrowAlignment}
  ${ArrowPlacement}
`
Popover.defaultProps = defaultProps

export default Popover
