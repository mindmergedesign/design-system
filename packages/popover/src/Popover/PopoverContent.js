import React, { Component } from 'react'
import ReactDOM from 'react-dom'
import { Popper } from 'react-popper'
import { Box, theme } from 'pcln-design-system'
import styled from 'styled-components'
import DEFAULTS_MODIFIERS from '../helpers/defaultModifiers'
import Overlay from '../overlay'
import PopoverArrow from '../arrow'

const defaultProps = {
  theme: theme,
  p: 2,
  bg: 'white',
  borderColor: 'borderGray',
  placement: 'top',
  zIndex: 102,
  width: 400
}

class PopoverContent extends Component {
  componentDidMount() {
    window.addEventListener('keyup', this.handleKeyUp, false)
  }

  componentWillUnmount() {
    window.removeEventListener('keyup', this.handleKeyUp, false)
  }

  handleKeyUp = evt => {
    const { onCloseRequest } = this.props
    const keys = {
      // Target ESC key
      27: () => {
        evt.preventDefault()
        onCloseRequest()
      }
    }

    if (keys[evt.keyCode]) {
      keys[evt.keyCode]()
    }
  }

  calcOffset = placement => {
    // Need to account for the padding added around the popover, this fixes the offset at the start and end position
    if (RegExp('start*').test(placement)) {
      return '-16px'
    } else if (RegExp('end*').test(placement)) {
      return '16px'
    } else {
      return 0
    }
  }

  render() {
    const { onCloseRequest } = this.props
    const styleProps = {
      theme: this.props.theme,
      background: this.props.bg,
      borderColor: this.props.borderColor,
      zIndex: this.props.zIndex,
      width: this.props.width
    }

    return ReactDOM.createPortal(
      <React.Fragment>
        <Popper
          positionFixed={true}
          modifiers={{
            ...DEFAULTS_MODIFIERS,
            offset: {
              offset: this.calcOffset(this.props.placement)
            }
          }}
          {...this.props}
        >
          {({ placement, ref, style, arrowProps }) => (
            // Need to be a native element, because of ref forwarding limitations with DS functional components
            <PopperGuide
              /*
               * NOTE: InnerRef has been depracted in V4 of styled components. Make sure to change this prop once we upgrade to styled components v4
               * https://www.styled-components.com/docs/api#deprecated-innerref-prop
               */
              innerRef={ref}
              style={style}
              data-placement={placement}
              aria-label={this.props.ariaLabel || 'Dialog Title'}
              {...styleProps}
              role="dialog"
              describedby={`dialog-description-${this.props.idx}`}
            >
              <ContentContainer
                innerRef={this.props.contentRef}
                {...styleProps}
                tabIndex="-1"
              >
                <Box id={`popover-description-${this.props.idx}`}>
                  {this.props.renderContent({
                    handleClose: onCloseRequest
                  })}
                </Box>
              </ContentContainer>
              <PopoverArrow
                arrowProps={arrowProps}
                placement={placement}
                background={styleProps.background}
                borderColor={styleProps.borderColor}
              />
            </PopperGuide>
          )}
        </Popper>
        <Overlay handleClick={onCloseRequest} zIndex={this.props.zIndex - 1} />
      </React.Fragment>,
      // Append each instance of the Popover as portal directly to the body
      document.querySelector('body')
    )
  }
}

const PopperGuide = styled.div`
  padding: 16px;
  z-index: ${({ zIndex }) => zIndex}
  max-width: ${({ width }) => width}px;
  width: 100%;
  box-sizing: border-box;
`
const ContentContainer = styled.section`
  box-shadow: 0 0 0 1px ${({ theme, borderColor }) => theme.colors[borderColor]},
    0 0 4px 0 rgba(0, 0, 0, 0.08), 0 8px 8px 0 rgba(0, 0, 0, 0.08),
    0 16px 16px 0 rgba(0, 0, 0, 0.08);
  font-size: ${({ theme }) => theme.fontSizes[0]}px;
  border-radius: ${({ theme }) => theme.radii[1]}px;
  background: ${({ theme, background }) => theme.colors[background]};
  box-sizing: border-box;
  outline: 0;
  max-width: 100%;
`

PopoverContent.defaultProps = defaultProps

export default PopoverContent
