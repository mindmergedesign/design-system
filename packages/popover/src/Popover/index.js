import Popover from './PopoverController'
export default Popover
