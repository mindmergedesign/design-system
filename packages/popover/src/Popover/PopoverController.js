import React, { Component } from 'react'
import { Manager, Reference } from 'react-popper'
import PopoverContent from './PopoverContent'

class PopoverController extends Component {
  contentRef = React.createRef()
  triggerRef = React.createRef()
  state = {
    isPopoverOpen: false
  }

  handleToggle = isOpen => {
    if (isOpen) {
      this.handleClose()
    } else {
      this.handleOpen()
    }
  }

  handleClose = () => {
    this.setState(
      prevState => {
        if (prevState.isPopoverOpen) {
          return {
            isPopoverOpen: false
          }
        }
      },
      () => {
        this.setFocusToRef(this.triggerRef)
      }
    )
  }

  handleOpen = () => {
    this.setState(
      prevState => {
        if (!prevState.isPopoverOpen) {
          return {
            isPopoverOpen: true
          }
        }
      },
      () => {
        this.setFocusToRef(this.contentRef)
      }
    )
  }

  setFocusToRef = ref => {
    if (ref) {
      try {
        ref.current.focus()
      } catch {
        return false
      }
    }
  }

  render() {
    const { isPopoverOpen } = this.state

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            // Need to be a native element, because of ref forwarding limitations with DS functional components
            <div ref={ref}>
              {// Clone element to pass down toggle event so it can be used directly from children as needed
              React.cloneElement(this.props.children, {
                'aria-label': 'Click to open popover with more information',
                onClick: () => this.handleToggle(isPopoverOpen),
                ref: this.triggerRef //Currently ref only works with native element, if we use a DS core component it does not work.
              })}
            </div>
          )}
        </Reference>
        {isPopoverOpen && (
          <PopoverContent
            {...this.props}
            onCloseRequest={this.handleClose}
            contentRef={this.contentRef}
          />
        )}
      </Manager>
    )
  }
}

export default PopoverController
