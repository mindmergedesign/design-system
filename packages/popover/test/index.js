import React from 'react'
import { render, fireEvent, wait, getByRole } from 'react-testing-library'
import Popover from '../src/popover'

const InnerContent = ({ handleClose }) => (
  <div>
    <p>Hello there!</p>
    <button onClick={handleClose}>Click me to close button!</button>
  </div>
)

const popoverProps = {
  renderContent: InnerContent,
  ariaLabel: 'Top PopOver',
  placement: 'auto',
  idx: 1,
  width: 400
}

const triggerButtonText = 'Trigger Button'

const { rerender, getByText, queryByText, container, debug, unmount } = render(
  <Popover {...popoverProps}>
    <button>{triggerButtonText}</button>
  </Popover>
)

describe('Popover', () => {
  describe('Trigger Element', () => {
    test('Renders trigger element and appends action props', () => {
      rerender(
        <Popover {...popoverProps}>
          <button>{triggerButtonText}</button>
        </Popover>
      )
      expect(container.firstChild).toMatchSnapshot()
      unmount()
    })

    test('clicking on trigger element, opens popover', () => {
      rerender(
        <Popover {...popoverProps}>
          <button>{triggerButtonText}</button>
        </Popover>
      )
      fireEvent.click(getByText(triggerButtonText))
      expect(getByText('Hello there!')).toBeTruthy()
      unmount()
    })

    test('clicking on close element inside popover, closes popover', () => {
      rerender(
        <Popover {...popoverProps}>
          <button>{triggerButtonText}</button>
        </Popover>
      )
      fireEvent.click(getByText(triggerButtonText))
      fireEvent.click(getByText('Click me to close button!'))
      expect(queryByText('Hello there!')).toBeFalsy()
      unmount()
    })
  })

  describe('UI Positioning', () => {
    test('Content is rendered according to placement prop', async () => {
      rerender(
        <Popover {...popoverProps} placement="bottom">
          <button>{triggerButtonText}</button>
        </Popover>
      )
      await wait(() => fireEvent.click(getByText(triggerButtonText)))
      expect(getByRole(document.body, 'dialog')).toMatchSnapshot()
    })
  })
})
