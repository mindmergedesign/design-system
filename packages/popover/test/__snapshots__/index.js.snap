// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Trigger Element Renders trigger element and appends action props 1`] = `
.c0 {
  display: inline-block;
}

<div
  class="c0"
>
  <button
    aria-label="Click to open popover with more information"
  >
    Trigger Button
  </button>
</div>
`;

exports[`Popover UI Positioning Content is rendered according to placement prop 1`] = `
.c2 {
  position: absolute;
}

.c2[data-placement*="right"] {
  left: 0;
}

.c2[data-placement*="left"] {
  right: 0;
}

.c2[data-placement*="top"] {
  bottom: 0;
}

.c2[data-placement*="bottom"] {
  top: 0;
}

.c2[data-placement*="bottom"]::before,
.c2[data-placement*="right"]::before,
.c2[data-placement*="top"]::after,
.c2[data-placement*="left"]::after {
  content: '';
  position: absolute;
  width: 0;
  height: 0;
  margin: auto;
  border-style: solid;
  -webkit-transform-origin: 0 0;
  -ms-transform-origin: 0 0;
  transform-origin: 0 0;
  -webkit-transform: rotate(-45deg);
  -ms-transform: rotate(-45deg);
  transform: rotate(-45deg);
  border-width: 6px;
}

.c2[data-placement*="right"]::before {
  left: 8px;
  border-color: #fff transparent transparent #fff;
  box-shadow: -0.75px -0.75px 0px 0.25px #c0cad5;
}

.c2[data-placement*="left"]::after {
  right: 13px;
  border-color: transparent #fff #fff transparent;
  box-shadow: 0.75px 0.75px 0px 0.25px #c0cad5;
}

.c2[data-placement*="top"]::after {
  top: -16px;
  margin-left: -5px;
  border-color: transparent transparent #fff #fff;
  box-shadow: -0.75px 0.75px 0px 0.25px #c0cad5;
}

.c2[data-placement*="bottom"]::before {
  top: 16px;
  margin-left: -5px;
  border-color: #fff #fff transparent transparent;
  box-shadow: 0.75px -0.75px 0px 0.25px #c0cad5;
}

.c0 {
  padding: 16px;
  z-index: 102;
  max-width: 400px;
  width: 100%;
  box-sizing: border-box;
}

.c1 {
  box-shadow: 0 0 0 1px #c0cad5, 0 0 4px 0 rgba(0,0,0,0.08),0 8px 8px 0 rgba(0,0,0,0.08), 0 16px 16px 0 rgba(0,0,0,0.08);
  font-size: 12px;
  border-radius: 2px;
  background: #fff;
  box-sizing: border-box;
  outline: 0;
  max-width: 100%;
}

<div
  aria-describedby="dialog-description-1"
  aria-label="Top PopOver"
  class="c0"
  data-placement="bottom"
  role="dialog"
  style="position: fixed; top: 0px; left: 0px; transform: translate3d(NaNpx, NaNpx, 0); will-change: transform;"
  width="400"
>
  <section
    class="c1"
    tabindex="-1"
    width="400"
  >
    <div
      class=""
      id="popover-description-1"
    >
      <div>
        <p>
          Hello there!
        </p>
        <button>
          Click me to close button!
        </button>
      </div>
    </div>
  </section>
  <span
    aria-hidden="true"
    class="c2"
    data-placement="bottom"
  />
</div>
`;
